### PR TITLE
[RFC] add ability to merge structured configs

### PR DESCRIPTION
**Pull request by teto**
POC

###### Motivation for this change
Listed here https://github.com/NixOS/nixpkgs/issues/41103. 
To sum up, I would like programs that require kernel modifications to be able to specify their kernel requirements. This might be disabled etc. but here is a proposition to show how it could look like

###### Things done

- [x] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [x] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nox --run "nox-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

---

follow up of https://github.com/NixOS/nixpkgs/pull/41393

### CODE REVIEW

**Comment by teto** [comment]
@dezgeg thanks for reviewing/merging the structured config PR \o/

Being able to specify required config for packages was one of the motivation. I would love some feedback on how to best achieve this since the current way is not practical aka:
```
openvswitch = stdenv.mkDerivation {
....
  passthru.kernelExtraConfig = with import ../../../../lib/kernel.nix { inherit (kernel) version; lib = stdenv.lib; }; {
    OPENVSWITCH  = whenAtLeast "2.0" yes;
  };
}
```
I moved the version dependant filters `whenAtLeast` etc.. out of lib/kernel.nix so that lib/kernel.nix can be imported by lib/default.nix. This replaces `with import ../../../../lib/kernel.nix { inherit (kernel) version; lib = stdenv.lib; };` with `with import lib.kernel;` but I am still stuck with
`  passthru.kernelExtraConfig = { OPENVSWITCH  = yes; }`

The problem is that if we just pass it like this
```
  my_lenovo_kernel = prev.linux_latest.override({
    structuredExtraConfig = prev.pkgs.openvswitch.passthru.kernelExtraConfig;
})
```
it erases conditions (e.g.,  `whenAtLeast` ) in common-config.nix such as 
`OPENVSWITCH  = optional whenAtLeast "3.4" yes;`

 I wonder if it wouldn't be better to have packages define a list of config items like `passthru.kernelExtraConfig  = [ OPENVSWITCH ];`and have the kernel config logic enable these the common-config.nix (strip the optional for instance) ?

To sum up, the best interface would be IMO to have in common-config.nix
`OPENVSWITCH  = optional whenAtLeast "3.4" no;` 
and find some way to turn this into `OPENVSWITCH  = whenAtLeast "3.4" yes;` when building my_lenovo_kernel (cf previous example).

**Comment by teto** [comment]
I am experimenting a bit as I am trying to solve the initial issue. Without types, the code is hard to follow but  what I've come up with is something similar to the module system so that hopefully merging 
`{ DEBUG = option yes; }` and  `{ DEBUG =  yes; }` returns sthg like `{ DEBUG = option yes; }`.
So for instance yes is defined as `yes = { answer = "y"; }`. My fear is that I am recreating the module system so any hindsight would be nice.

File: lib/debug.nix [hunk 1]
```diff
@@ -77,7 +77,7 @@ rec {
                (modify depth snip x)) y;
 
   /* A combination of `traceVal` and `traceSeq` */
-  traceValSeqFn = f: v: traceVal f (builtins.deepSeq v v);
+  traceValSeqFn = f: v: traceValFn f (builtins.deepSeq v v);
```
Comment (Profpatsch) [inline lib/debug.nix]: Huh, I thought we merged that line already?

Comment (teto) [reply lib/debug.nix]: it's not up to date. I've started modifying the code to adopt somewhat the module backend so it might be best not to review it now

**Comment by teto** [comment]
I went ahead and proceeded with the modifications:

- extraStructuredConfig now expects a list of attributes. It makes the composition of the kernel config easier.
- I've proceeded with distinguishing freeform options from tristate ones
- will look for a structured config in kernelPatches too
- one can now access the structuredConfig from a kernel via linux_test.configfile.structuredConfig in order to reinject it another kernel, no need to rewrite the config from scratch
- I've defined the following merge strategies in case of conflict:
-- freeform items must be equal or they conflict (mergeEqualOption)
-- for tristate (y/m/n) entries, I use the mergeAnswer strategy which takes the best available value, "best" being defined by the user (by default "y" > "m" > "n", e.g. if one entry is both marked "y" and "n", "y" wins)
-- if one item is both marked optional/mandatory, mandatory wins (mergeFalseByDefault)


Here is what I test with:

```
    linux_test = let 
      configStructured = with prev.lib.kernel; [ 


        # common-config.nix default is 32 

        { MMC_BLOCK_MINORS   = freeform "32"; } # same as default, won't trigger any error 
        # { MMC_BLOCK_MINORS   = freeform "64"; } # will trigger an error but the message is not great:
        # The option `settings.MMC_BLOCK_MINORS.freeform' has conflicting definitions, in `<unknown-file>' and `<unknown-file>'

        # mandatory should win by default
        { USB_DEBUG = option yes;}
        # { USB_DEBUG = yes;}

        # default for "8139TOO_PIO" is no
        { "8139TOO_PIO"  = yes; } # yes wins over no by default
      ];
    in
    prev.linux_latest.override {
      kernelPreferBuiltin = true;
      structuredExtraConfig = configStructured;
  };
```

I am sure the code may be improved and there are some cases that might break. I would like to write some tests but not sure how to do.

**Comment by edolstra** [comment]
Can `lib/kernel.nix` be moved out of `lib`? `lib` really shouldn't contain package-specific functions.

**Comment by dezgeg** [comment]
The kconfig language/mechanism is actually quite popular in the embedded community, for instance Buildroot, BusyBox, Crosstool-ng, OpenWRT, U-Boot all use it. Nothing of those that's currently packaged needs to care about the config file to the same precision as our kernel package does, however.

**Comment by teto** [comment]
I would like packages to be able to specify kernel configurations, like enabling openvswitch should set `OPENVSWITCH = yes; `. To make yes/no available to packages I put them into lib/kernel.nix. 
I can indeed move the rest ouf of lib/kernel.nix.

**Comment by teto** [comment]
I fixed the only trouble I had left when converting the optional configurations (some dropped optional items). I was able to compile 4_4/4_9/4_18/linux_hardened.

@edolstra I left only some aliases in https://github.com/NixOS/nixpkgs/blob/826e270c8d49f8876e48583c3f70672c4d250b11/lib/kernel.nix in kernel.nix for modules/packages to configure the kernel. But they could do without it and use the full config { optional = false; tristate = "y"; } instead so I can get rid of lib/kernel.nix entirely if you insist.

My only concern now is that I changed structuredExtraConfig from being a set (current master)  to a list of set (so that lib.evalModules merges the configs). I wonder if instead one could not just pass an structuredExtraConfig=mkMerge[ {} {} ] to get the same results ?

**Comment by teto** [comment]
1/ I wanted to add tests for the strategies described in https://github.com/NixOS/nixpkgs/pull/42838#issuecomment-411999672 . Confirming these are the default strategies we want would help.
2/ I want to compose kernel configs (kvmConfig + ebpfConfig + mptcpConfig etc). This is already possible via passing extraStructuredConfig=[ kvmConfig ebpfConfig mptcpConfig ] but I would prefer  extraStructuredConfig=mkMerge[ kvmConfig  ebpfConfig mptcpConfig] but I haven't had the time to test it yet.
3/ I wanted to revert the conversion of mptcp and hardened configs to slim the PR a bit.

**Comment by dtzWill** [comment]
Okay! Looking forward to this!

**Comment by teto** [comment]
I've addressed all the points in my previous message https://github.com/NixOS/nixpkgs/pull/42838#issuecomment-419328767

From my side the PR is feature complete so now it's up to reviewers to decide what's wrong/false 
To run the tests `$ nix-build -A tests.kernel-config  ~/nixpkgs3`

A kernel I have in my overlay to test
```
    linux_test = let 
      # mininetConfigStructured = {};
      configStructured = with prev.lib.kernel; prev.lib.mkMerge [ 
    BPF = yes;
    BPF_SYSCALL = yes;
    VIRTIO            = mkForce yes;


      ];
    in
    prev.linux_latest.override {
      kernelPreferBuiltin = true;
      structuredExtraConfig = configStructured;
  };
```

**Comment by teto** [comment]
Needed to test master version of alot with python3 support @Fuuzetsu @primeos

**Comment by teto** [comment]
bump ?

File: pkgs/os-specific/linux/kernel/common-config.nix [hunk 2]
```diff
@@ -12,25 +12,19 @@
...
-assert (mkValueOverride == null) || (builtins.isFunction mkValueOverride);
-
 with stdenv.lib;
 
-with import ../../../../lib/kernel.nix { inherit (stdenv) lib; inherit version; };
+  with import ../../../../lib/kernel.nix { inherit (stdenv) lib; };
 
 let
+  # Common patterns/legacy
+  when = cond: opt: if cond then opt else null;
```
Comment (danbst) [inline pkgs/os-specific/linux/kernel/common-config.nix]: why isn't this `mkIf` too?

**Review by infinisil — CHANGES_REQUESTED** [review]

File: pkgs/os-specific/linux/kernel/generic.nix [hunk 3]
```diff
@@ -130,7 +135,30 @@ let
...
+
+      module = import ../../../../nixos/modules/system/boot/kernel_config.nix;
+      # used also in apache
+      # { modules = [ { options = res.options; config = svc.config or svc; } ];
+      #   check = false;
+      # The result is a set of two attributes
+      moduleStructuredConfig = (lib.evalModules {
+        modules = [
+          module
+          { settings = commonStructuredConfig; }
```
[the remark this answers is not in the document]
Comment (teto) [reply pkgs/os-specific/linux/kernel/generic.nix]: done

File: nixos/modules/system/boot/kernel_config.nix [hunk 4]
```diff
@@ -0,0 +1,137 @@
...
+    mkConf = cfg: concatStrings (mapAttrsToList mkConfigLine cfg);
+  in mkConf exprs;
+
+in
+{
+
+  options = {
+
+    intermediateNixConfig = mkOption {
+      readOnly = true;
```
[the remark this answers is not in the document]
Comment (teto) [reply nixos/modules/system/boot/kernel_config.nix]: > This has nothing to do with Nix after all

Yet the format is a nix one, only understood by the nix perl script. Yet this might be unnecessary precise and configString is ok too. 
Regarding the readOnly parameter, even if I change it, users won't be able to set intermediateNixConfig as this PR does not expose the module. It's just used internally to convert a structured set into intermediateNixConfig.
Thus right now the PR should not break users setup. The old `extraConfig` will keep working and if they wish, they can use the new `structuredExtraConfig. I agree with your vision but I think it's best kept for a future PR, let's keep this one small.

[the remark this answers is not in the document]
Comment (teto) [reply nixos/modules/system/boot/kernel_config.nix]: renaming `settings` to `config` breaks something here and it also makes the reading a bit harder with things like `config.config` . I would prefer to name it `config` too but as this is not exposed to users yet (they will pass their config via `structuredExtraConfig`), it can be renamed later without breaking.

File: nixos/modules/system/boot/kernel_config.nix [hunk 5]
```diff
@@ -0,0 +1,137 @@
...
+        MMC_BLOCK_MINORS = freeform "32";
+      }'';
+      description = ''
+        Structured kernel configuration.
+      '';
+    };
+  };
+
+  config = {
+    intermediateNixConfig = generateNixKConf config.settings;
```
[the remark this answers is not in the document]
Comment (teto) [reply nixos/modules/system/boot/kernel_config.nix]: see my previous comment.

File: pkgs/os-specific/linux/kernel/generic.nix (hunk 3 above)
[the remark this answers is not in the document]
Comment (teto) [reply pkgs/os-specific/linux/kernel/generic.nix]: I believe it should be hard to override as it contains the default nixos config from common-config.nix. This kind of consideration will make more sense when/if we expose a boot.kernel.structuredConfig I suppose.

File: nixos/modules/system/boot/kernel_config.nix (hunk 4 above)
[the remark this answers is not in the document]
Comment (Infinisil) [reply nixos/modules/system/boot/kernel_config.nix]: Ah I see, didn't know this wasn't exposed. All good then.

**Review by infinisil — APPROVED** [review]
This looks good to me for the code side, I haven't tested any of this though, but I have trust in @teto to have done this. Assuming all kernel options are the same as before, this can hardly break anything anyways and is mostly a cosmetic change.

Thanks for all the work @teto! If nobody has any complaints, I'll merge this when I see it next time.

**Comment by grahamc** [comment]
@grahamcofborg eval

**Comment by teto** [comment]
wait a bit While I had made sure configs were identical in the initial PR, after the rebase: I 've just run
`nix-build -A linux_latest_hardened.configfile  ~/nixpkgs` and `nix-build -A linux_latest_hardened.configfile  ~/nixpkgs3` (master and PR) and seen some differences. Working on it.

**Comment by Infinisil** [comment]
@teto I see, let me know when you're certain

**Comment by teto** [comment]
I did well to check several mistakes had crept in.
The diff between hardened kernels now returns nothing \o/ (I just checked for linux_hardened).
The lib/kernels.nix isn't made public anymore and is imported just by the files needing it. 
To sum it up, the PR exposes nothing new (structuredExtraConfig is already available in nixos-unstable), it just adds an internal mechanism to merge kernel configurations.
Once this PR is merged, we need to think about what and how to expose kernel configuration to packages / user (boot.kernelConfiguration etc, might warrant an RFC).

**Comment by Infinisil** [comment]
@teto Merge conflict

**Comment by teto** [comment]
@Infinisil ty, pushed an update.

**Comment by Infinisil** [comment]
Evaluation fails now, messed up in all-packages.nix

**Comment by dtzWill** [comment]
hype hype, thanks all! :D

**Comment by teto** [comment]
@Infinisil sorry, the lights were green after my last push, I guess some caching problem on the browser. The timeout for the first check seems to appear once the config was generated so I think it's fine. 

@dtzWill very excited about this too. The current achievement is the possibility to do `extraStructuredConfig=mkMerge [ ]` and be warned in case of contradicting configurations. Yet it allows for some exciting future possibilities.

**Comment by teto** [comment]
you are right, it shouldn't sorry for that, let me look into this.

**Comment by teto** [comment]
@lopsided98 what kind of error do you have ? I have
```
make: Entering directory '/build/source'
make[1]: Entering directory '/build/source/build'
  HOSTCC  scripts/basic/fixdep
  GEN     ./Makefile
  HOSTCC  scripts/kconfig/conf.o
  SHIPPED scripts/kconfig/zconf.tab.c
  SHIPPED scripts/kconfig/zconf.lex.c
  HOSTCC  scripts/kconfig/zconf.tab.o
  HOSTLD  scripts/kconfig/conf
***
*** Can't find default configuration "arch/x86/configs/odroidxu4_defconfig"!
***

```
which shouldn't be influenceed by my changes. Also I can reproduce this error on nixos-unstable so it must be some other change
nix build -f channel:nixos-unstable linux_hardkernel_4_14

**Comment by lopsided98** [comment]
Your error is caused by building it on x86; `odroidxu4_defconfig` only works on ARM. You should be able to cross compile it with `nix build -f channel:nixos-unstable pkgsCross.armv7l-hf-multiplatform.linux_hardkernel_4_14`.

My error is:
```
<3>  CC [M]  drivers/gator/gator_events_mali_midgard.o
<3>../drivers/gator/gator_events_mali_midgard.c:25:10: fatal error: linux/mali_linux_trace.h: No such file or directory
<3> #include "linux/mali_linux_trace.h"
<3>          ^~~~~~~~~~~~~~~~~~~~~~~~~~
<3>compilation terminated.
<3>make[4]: *** [../scripts/Makefile.build:326: drivers/gator/gator_events_mali_midgard.o] Error 1
<3>make[3]: *** [../scripts/Makefile.build:585: drivers/gator] Error 2
<3>make[2]: *** [/build/source/Makefile:1044: drivers] Error 2
<3>make[2]: *** Waiting for unfinished jobs....
```

This error was fixed in 5ac73349506894059871d94b7418e411c812f717 by adding `GATOR n` to `extraConfig`, so the return of this error implies that `extraConfig` is no longer being applied.

**Comment by teto** [comment]
seems like you are right, extraConfig is not taken into account. Does
```
diff --git a/pkgs/os-specific/linux/kernel/generic.nix b/pkgs/os-specific/linux/kernel/generic.nix
index a41f1eb989b..7cd134805f1 100644
--- a/pkgs/os-specific/linux/kernel/generic.nix
+++ b/pkgs/os-specific/linux/kernel/generic.nix
@@ -76,7 +76,7 @@ let
   # extra config in legacy string format
   extraConfig = extraConfig + lib.optionalString (stdenv.hostPlatform.platform ? kernelExtraConfig) stdenv.hostPlatform.platform.kernelExtraConfig;
 
-  intermediateNixConfig = configfile.moduleStructuredConfig.intermediateNixConfig;
+  intermediateNixConfig = configfile.moduleStructuredConfig.intermediateNixConfig + extraConfig;
 
   structuredConfigFromPatches =
         map ({extraStructuredConfig ? {}, ...}: {settings=extraStructuredConfig;}) kernelPatches;

```
this solve it ?

**Comment by lopsided98** [comment]
```
error: while evaluating the attribute 'drvPath' at /home/ben/Documents/Projects/nixpkgs/lib/customisation.nix:149:7:
while evaluating the attribute 'configurePhase' of the derivation 'linux-4.14.87-153-armv7l-unknown-linux-gnueabihf' at /home/ben/Documents/Projects/nixpkgs/pkgs/stdenv/generic/make-derivation.nix:183:11:
while evaluating the attribute 'kernelConfig' of the derivation 'linux-config-4.14.87-153-armv7l-unknown-linux-gnueabihf' at /home/ben/Documents/Projects/nixpkgs/pkgs/stdenv/generic/make-derivation.nix:183:11:
while evaluating 'kernelConfigFun' at /home/ben/Documents/Projects/nixpkgs/pkgs/os-specific/linux/kernel/generic.nix:85:21, called from /home/ben/Documents/Projects/nixpkgs/pkgs/os-specific/linux/kernel/generic.nix:97:20:
infinite recursion encountered, at /home/ben/Documents/Projects/nixpkgs/pkgs/os-specific/linux/kernel/generic.nix:77:17
```

**Comment by teto** [comment]
I've updated the code

**Comment by lopsided98** [comment]
That patch fixes the issue.

File: nixos/modules/system/boot/kernel_config.nix [hunk 6]
```diff
@@ -0,0 +1,137 @@
...
+      description = ''
+        The result of converting the structured kernel configuration in settings
+        to an intermediate string that can be parsed by generate-config.pl to
+        answer the kernel `make defconfig`.
+      '';
+    };
+
+    settings = mkOption {
+      type = types.attrsOf kernelItem;
+      example = literalExample '' with lib.kernel; {
```
Comment (danbst) [inline nixos/modules/system/boot/kernel_config.nix]: @teto Note that `with lib.kernel;` is oversimplification. There is no, top-level `kernel` attribute in `lib`.
What is really wanted here, is like
```
  with import ../../../../lib/kernel.nix { inherit (stdenv) lib; inherit version; };
```
Maybe make kernel lib an attribute in `linuxPackages`? My usecase is this (self-contained NixOS config):
```
{ pkgs, lib, ... }: {
  boot.kernelPackages = pkgs.linuxPackages_latest.extend (lib.const (ksuper: {
    kernel =
      with import <nixpkgs/lib/kernel.nix> { inherit lib; version = ksuper.kernel.version; };
      ksuper.kernel.override {
        structuredExtraConfig.SND = no;
      };
  }));
  boot.loader.grub.device = "/nodev";
  fileSystems."/".device = "/nodev";
}
```
And I imagine it like this;
```
{ pkgs, lib, ... }: {
  boot.kernelPackages = pkgs.linuxPackages_latest.extend (lib.const (ksuper: {
    kernel =
      with ksuper.lib; # or even ksuper.kernel.lib
      ksuper.kernel.override {
        structuredExtraConfig.SND = no;
      };
  }));
}
```

Comment (teto) [reply nixos/modules/system/boot/kernel_config.nix]: The description hasn t been updated after lib.kernel got removed. The reason for removal was that i wanted to propose a more holistic version before committing to a public api. So right now structured configs are not practical to use unless you know the implementation (i.e., your solution). I 've started drafting a proposal but i got kinda busy. The idea being to solve https://github.com/NixOS/nixpkgs/issues/41103

Comment (danbst) [reply nixos/modules/system/boot/kernel_config.nix]: Ah, maybe make it back?
```
{ lib, version ? null }:

with lib;
rec {
  # Common patterns/legacy
  whenAtLeast = ver: mkIf (versionAtLeast version ver);
  whenOlder   = ver: mkIf (versionOlder version ver);
  # range is (inclusive, exclusive)
  whenBetween = verLow: verHigh: mkIf (versionAtLeast version verLow && versionOlder version verHigh);
 ...
}
```
Then resolve "null" version as "current version" in merger. Even without `whenBetween`, `whenOlder` and `whenAtLeast` this `lib.kernel` is useful (see https://github.com/NixOS/nixpkgs/issues/55415)

**Comment by teto** [comment]
Elco didn t want this to leak. Instead of using the alias 'no' you can also weite the submodule yourself: {answer="n";}
